src/recluster.c

```diff
@@ -0,0 +1,1175 @@
+		CatalogTupleUpdate(relRelation, &reltup->t_self, reltup);
+	else
+		CacheInvalidateRelcacheByTuple(reltup);
+
+	/* Clean up. */
+	heap_freetuple(reltup);
+	heap_close(relRelation, RowExclusiveLock);
+
+	/* Make the update visible */
+	CommandCounterIncrement();
+}
+
+
+
+/*
+ * Remove the transient table that was built by make_new_heap, and finish
+ * cleaning up (including rebuilding all indexes on the old heap).
+ *
+ * NB: new_index_oids must be in the same order as RelationGetIndexList
+ *
+ */
+static void
+finish_heap_swaps(Oid OIDOldHeap, Oid OIDNewHeap,
+				  List *new_index_oids,
+				  bool swap_toast_by_content,
+				  bool is_internal,
+				  TransactionId frozenXid,
+				  MultiXactId cutoffMulti)
+{
+	ObjectAddress object;
+	Relation	oldHeapRel;
+	List	   *old_index_oids;
+	ListCell   *old_index_cell;
+	ListCell   *new_index_cell;
+
+
+	/*
+	 * There's a risk of deadlock if some other process is also trying to
+	 * upgrade their lock in the same manner as us, at this time. Since our
+	 * transaction has performed a large amount of work, and only needs to be
+	 * run once per chunk, we do not want to abort it due to this deadlock. To
+	 * prevent abort we set our `deadlock_timeout` to a large value in the
+	 * expectation that the other process will timeout and abort first.
+	 * Currently we set `deadlock_timeout` to 1 hour, as this should be longer
+	 * than any other normal process, while still allowing the system to make
+	 * progress in the event of a real deadlock. As this is the last lock we
+	 * grab, and the setting is local to our transaction we do not bother
+	 * changing the guc back.
+	 */
+	/* TODO is this the right timeout value? */
+	int			config_change = set_config_option("deadlock_timeout",
+												  "101000",
+												  PGC_SUSET,
+												  PGC_S_SESSION,
+												  GUC_ACTION_LOCAL,
+												  true, 0, false);
+
+	if (config_change == 0)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("deadlock_timeout guc does not exist.")));
+	else if (config_change < 0)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("could not set deadlock_timeout guc.")));
+
+	oldHeapRel = heap_open(OIDOldHeap, AccessExclusiveLock);
+
+	/*
+	 * Swap the contents of the heap relations (including any toast tables).
+	 * Also set old heap's relfrozenxid to frozenXid.
+	 */
+	swap_relation_files(OIDOldHeap, OIDNewHeap,
+						swap_toast_by_content, is_internal,
+						frozenXid, cutoffMulti);
+
+	/* Swap the contents of the indexes */
+	old_index_oids = RelationGetIndexList(oldHeapRel);
+	forboth(old_index_cell, old_index_oids, new_index_cell, new_index_oids)
+	{
+		Oid			old_index_oid = lfirst_oid(old_index_cell);
+		Oid			new_index_oid = lfirst_oid(new_index_cell);
+
+		swap_relation_files(old_index_oid, new_index_oid,
+		/* TODO */
+							swap_toast_by_content, true,
+							frozenXid, cutoffMulti);
+	}
+	/* TODO assert same length? */
```

Comment:
`RelationGetIndexList` guarantees that the list is sorted in order by OID, and since we we're locking out writes the set of indices cannot be changed, so it should be safe, but I agree it'll be nicer to just return both lists.